src/manager/evm/manager.rs

```diff
@@ -533,36 +560,97 @@ impl<M: Middleware + Send + Sync + 'static> EthManager for EthSubnetManager<M> {
     pub fn new(
-        eth_client: Arc<M>,
-        gateway_contract: Gateway<Arc<M>>,
-        registry_contract: SubnetRegistry<Arc<M>>,
+        gateway_addr: ethers::types::Address,
+        registry_addr: ethers::types::Address,
+        chain_id: u64,
+        provider: Provider<Http>,
+        keystore: Arc<RwLock<PersistentKeyStore<ethers::types::Address>>>,
     ) -> Self {
         Self {
-            eth_client,
-            gateway_contract,
-            registry_contract,
+            keystore,
+            on_chain_info: OnChainInfo {
+                gateway_addr,
+                registry_addr,
+                chain_id,
+                provider,
+            },
         }
     }
 
-    pub fn ensure_same_gateway(&self, gateway: &Address) -> anyhow::Result<()> {
+    pub fn ensure_same_gateway(&self, gateway: &Address) -> Result<()> {
         let evm_gateway_addr = payload_to_evm_address(gateway.payload())?;
-        if evm_gateway_addr != (*self.gateway_contract).address() {
+        if evm_gateway_addr != self.on_chain_info.gateway_addr {
             Err(anyhow!("Gateway address not matching with config"))
         } else {
             Ok(())
         }
     }
+
+    fn get_registry_instance(
+        &self,
+        signer_addr: Option<&ethers::types::Address>,
+    ) -> Result<SubnetRegistry<MiddlewareImpl>> {
+        Ok(SubnetRegistry::new(
+            self.on_chain_info.registry_addr,
+            Arc::new(self.get_ethers_client(signer_addr)?),
+        ))
+    }
+
+    fn get_gateway_instance(
+        &self,
+        signer_addr: Option<&ethers::types::Address>,
+    ) -> Result<Gateway<MiddlewareImpl>> {
+        Ok(Gateway::new(
+            self.on_chain_info.gateway_addr,
+            Arc::new(self.get_ethers_client(signer_addr)?),
+        ))
+    }
+
+    fn get_subnet_contract_instance(
+        &self,
+        contract_addr: ethers::types::Address,
+        signer_addr: Option<&ethers::types::Address>,
+    ) -> Result<SubnetContract<MiddlewareImpl>> {
+        Ok(SubnetContract::new(
+            contract_addr,
+            Arc::new(self.get_ethers_client(signer_addr)?),
+        ))
+    }
+
+    /// Get the ethers client instance. If addr is None, we expect the client to call only `view`
+    /// functions. To do that, we use a random generated byte private key (dont send any funds to it)
+    /// so that the interface is consistent.
+    fn get_ethers_client(&self, addr: Option<&ethers::types::Address>) -> Result<MiddlewareImpl> {
+        let wallet = if let Some(addr) = addr {
+            let keystore = self.keystore.read().unwrap();
+            let private_key = keystore
+                .get(addr)?
+                .ok_or_else(|| anyhow!("address {addr:} does not have private key in key store"))?;
+            LocalWallet::from_bytes(private_key.private_key())?
+                .with_chain_id(self.on_chain_info.chain_id)
+        } else {
+            LocalWallet::from_bytes(&RANDOM_PRIVATE_KEY)?
```

Comment:
Is this the right way to do it? Can't we initialize here a `Middleware` instead of a `SignerMiddleware` to call the `view` functions? I feel generating a new random key every time can be really inefficient.

Reply:
This issue no longer present in latest code.